README
1. What's this?
===============

scim-prime is a SCIM IMEngine module using PRIME.

SCIM is Smart Common Input Method platform, it provides user friendly/full
featured user interface and development platform to make input method
developer's life easier.

PRIME is a predictive Japanese conversion engine written in Ruby.

scim-prime combines PRIME to SCIM and provides some usefull user interfaces.


2. Requirements
===============

 * scim-1.0.0 or later.
   - http://www.scim-im.org/

 * PRIME.
   - http://taiyaki.org/prime/index.html

3. Install
==========

You can build & install scim, prime and scim-prime by following command:

  $ ./configure
  $ make
  # make install

If you installed the SCIM into /usr/local or any other non-standard
directories, you should specify the PKG_CONFIG_PATH environment variable:

  $ PKG_CONFIG_PATH=/usr/local/lib/pkgconfig ./configure


4. Usage
========

Please run the SCIM with some environment variable from ~/.xinitrc or a same
kind file:

  export XMODIFIRES="@im=SCIM"
  export GTK_IM_MODULE=scim
  # specify QT_IM_MODULE if you have scim-qtimmm
  # export QT_IM_MODULE=scim
  scim &

After rebooting X Window system you'll find a keyboard icon in the system tray
and SCIM toolbar. You can switch input method to the scim-prime from these
items.


5. License
==========

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 59 Temple
Place - Suite 330, Boston, MA 02111-1307, USA.


6. Credit
=========

 * Takuro Ashie <ashie at homa.ne.jp>
   - Developer.

Special thanks to all contributers and users!
(Please see the ChangeLog to find all contributers.)


NEWS
Overview of Changes from scim-prime-1.0.0 to scim-prime-1.0.1
=============================================================
* Fix a build problem on gcc-4.3.

Overview of Changes from scim-prime-0.9.0 to scim-prime-1.0.0
=============================================================
* Fix credit.
* Version-1.0.0.

Overview of Changes from scim-prime-0.3.2 to scim-prime-0.9.0
=============================================================
* Add a preference for number of candidates in a page of the candidates window.
* Show version information at the setup window.

Overview of Changes from scim-prime-0.3.1 to scim-prime-0.3.2
=============================================================
* The problem that scim-prime seldom losts the connection to the PRIME on
  certain ennvironment has been Fixed.

Overview of Changes from scim-prime-0.3.0 to scim-prime-0.3.1
=============================================================
* Add "Control+J" for toggling ON/OFF PRIME.
* Add help text.
* Fix a clash bug of setup module.

Overview of Changes from scim-prime-0.2.10 to scim-prime-0.3.0
==============================================================
* Fix some wrong key bindings (Page up & Page down candidates).
* Fix a problem that scim-prime losts the connection to the PRIME process when
  the I/O is busy.
* Fix a crash problem of scim-setup when use scim-prime with glibc-2.3.5's
  iconv & scim-skk.
* Fix some english messages.
* Add SKIM support (provide as external package named skim-scim-prime).

Overview of Changes from scim-prime-0.2.9 to scim-prime-0.2.10
==============================================================
* Fix a bug that set wrong key bind to key bindings view when startup the
  scim-setup.

Overview of Changes from scim-prime-0.2.8 to scim-prime-0.2.9
=============================================================
* Fix clash bug of scim-setup. (Thanks James Su)
* Show all key bindings by default on scim-setup.
* Show detailed error message when fails to communicate with PRIME.
* Fix a wrong key binding. (Thanks Bando-san)

Overview of Changes from scim-prime-0.2.7 to scim-prime-0.2.8
=============================================================
* Add a key binding to toggle on/off PRIME.
* Add default language setting.
* Fix a build problem on non glibc system.

Overview of Changes from scim-prime-0.2.6 to scim-prime-0.2.7
=============================================================
* Colorize annotation, usage, comment strings on the lookup window.

Overview of Changes from scim-prime-0.2.5 to scim-prime-0.2.6
=============================================================
* Fix for gcc-4.0.
* Changed default setting of lookup window position to "tail" because "head"
  doesn't work correctly on certain environment such as scim-qtimm.

Overview of Changes from scim-prime-0.2.4 to scim-prime-0.2.5
=============================================================
* Update automake version because previous version reduce filename extension
  of the module.

Overview of Changes from scim-prime-0.2.3 to scim-prime-0.2.4
=============================================================
* Fix a bug that scim-prime eats all key events although it doesn't have
  preedit string (Thanks! DAIRIKI-san).
* Try recovery automatically when the PRIME connection is lost.
* Other minor bug fixes.

Overview of Changes from scim-prime-0.2.2 to scim-prime-0.2.3
=============================================================
* Add a preference for the position of the prediction window.
* Don't relay any key event to application while preediting.
* Fix typo in Japanese traslation. Thanks Kazuhiro NISHIYAMA.
* Other minor bug fixes. Thanks UTUMI Hirosi.

Overview of Changes from scim-prime-0.2.1 to scim-prime-0.2.2
=============================================================
* Refine key preference UI.
* Minor bug fixes.

Overview of Changes from scim-prime-0.2.0 to scim-prime-0.2.1
=============================================================
* Add an option to choose whether starts converting on inputting comma/period
  or not.
* Add an option to choose whether commits comma/period immediatly or not.
* Reverse color of preedit string on inline prediction mode.

Overview of Changes from scim-prime-0.1.0 to scim-prime-0.2.0
=============================================================
* Enable to edit sentence segments.
* Add inline prediction feature.
* Improve English mode.
* Add some key bindings.
* Fix some default settings.

scim-prime-0.1.0
=============================================================
* Initial release.


src/prime_commands.h
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 *  Copyright (C) 2005 Takuro Ashie <ashie@homa.ne.jp>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */


#ifndef __PRIME_COMMANDS_H__
#define __PRIME_COMMANDS_H__


#define PRIME_LANG_JAPANESE "Japanese"
#define PRIME_LANG_ENGLISH  "English"
#define PRIME_LANG_PRIME08  "Prime08"


#define PRIME_VERSION                "version"                // noargs

#define PRIME_REFRESH                "refresh"                // noargs
#define PRIME_CLOSE                  "close"                  // noargs

#define PRIME_SESSION_START          "session_start"          // arg1:LANGUAGE
#define PRIME_SESSION_END            "session_end"            // arg1:SESSION
#define PRIME_SESSION_GET_ENV        "session_get_env"        // arg1:SESSION
                                                              // arg2:VARIABLE

#define PRIME_LOOKUP                 "lookup"                 // arg1:PATTERN
#define PRIME_LOOKUP_ALL             "lookup_all"             // arg1:PATTERN
#define PRIME_LOOKUP_COMPACT         "lookup_compact"         // arg1:PATTERN
#define PRIME_LOOKUP_COMPACT_ALL     "lookup_compact_all"     // arg1:PATTERN
#define PRIME_LOOKUP_DIRECT          "lookup_direct"          // arg1:PATTERN
#define PRIME_LOOKUP_DIRECT_ALL      "lookup_direct_all"      // arg1:PATTERN
#define PRIME_LOOKUP_EXPANSION       "lookup_expansion"       // arg1:PATTERN
#define PRIME_LOOKUP_HYBRID          "lookup_hybrid"          // arg1:PATTERN
#define PRIME_LOOKUP_HYBRID_ALL      "lookup_hybrid_all"      // arg1:PATTERN
#define PRIME_LOOKUP_MIXED           "lookup_mixed"           // arg1:PATTERN
#define PRIME_LOOKUP_PREFIX          "lookup_prefix"          // arg1:PATTERN
#define PRIME_LOOKUP_PREFIX_EX       "lookup_prefix_ex"       // arg1:PATTERN

#define PRIME_EDIT_COMMIT            "edit_commit"            // arg1:SESSION
#define PRIME_EDIT_CURSOR_LEFT       "edit_cursor_left"       // arg1:SESSION
#define PRIME_EDIT_CURSOR_LEFT_EDGE  "edit_cursor_left_edge"  // arg1:SESSION
#define PRIME_EDIT_CURSOR_RIGHT      "edit_cursor_right"      // arg1:SESSION
#define PRIME_EDIT_CURSOR_RIGHT_EDGE "edit_cursor_right_edge" // arg1:SESSION
#define PRIME_EDIT_DELETE            "edit_delete"            // arg1:SESSION
#define PRIME_EDIT_ERASE             "edit_erase"             // arg1:SESSION
#define PRIME_EDIT_GET_QUERY_STRING  "edit_get_query_string"  // arg1:SESSION
#define PRIME_EDIT_GET_PREEDITION    "edit_get_preedition"    // arg1:SESSION
#define PRIME_EDIT_BACKSPACE         "edit_backspace"         // arg1:SESSION
#define PRIME_EDIT_INSERT            "edit_insert"            // arg1:SESSION
                                                              // arg2:STRING
#define PRIME_EDIT_SET_MODE          "edit_set_mode"          // arg1:SESSION
                                                              // arg2:MODE (default, katakana, half_katakana, wide_ascii, raw)
#define PRIME_EDIT_UNDO              "edit_undo"              // arg1:SESSION

#define PRIME_CONTEXT_RESET          "context_reset"          // arg1:SESSION
#define PRIME_CONTEXT_SET_PREVIOUS_WORD \
                                     "context_set_previous_word"
                                                              // arg1: SESSION
                                                              // arg2: WORD

#define PRIME_CONV_COMMIT            "conv_commit"            // arg1: SESSION
#define PRIME_CONV_CONVERT           "conv_convert"           // arg1: SESSION
                                                              // arg2: METHOD
#define PRIME_CONV_PREDICT           "conv_predict"           // arg1: SESSION
                                                              // arg2: METHOD
#define PRIME_CONV_SELECT            "conv_select"            // arg1: SESSION
                                                              // arg2: INDEX

#define PRIME_REGISTER_WORD          "register_word"          // arg1:SESSION
                                                              // arg2:READING
                                                              // arg3:LITERAL
                                                              // arg4:POS

#define PRIME_SEGMENT_COMMIT         "segment_commit"         // arg1:SESSION
#define PRIME_SEGMENT_RECONVERT      "segment_reconvert"      // arg1:SESSION
#define PRIME_SEGMENT_SELECT         "segment_select"         // arg1:SESSION
                                                              // arg1:INDEX

#define PRIME_MODIFY_START              "modify_start"
#define PRIME_MODIFY_CURSOR_EXPAND      "modify_cursor_expand"
#define PRIME_MODIFY_CURSOR_SHRINK      "modify_cursor_shrink"
#define PRIME_MODIFY_CURSOR_LEFT        "modify_cursor_left"
#define PRIME_MODIFY_CURSOR_LEFT_EDGE   "modify_cursor_left_edge"
#define PRIME_MODIFY_CURSOR_RIGHT       "modify_cursor_right"
#define PRIME_MODIFY_CURSOR_RIGHT_EDGE  "modify_cursor_right_edge"
#define PRIME_MODIFY_GET_CANDIDATES     "modify_get_candidates"
#define PRIME_MODIFY_GET_CONVERSION     "modify_get_conversion"

#define PRIME_LEARN_WORD             "learn_word"             // arg1:KEY
                                                              // arg2:VALUE
                                                              // arg3:PART
                                                              // arg4:CONTEXT
                                                              // arg5:SUFFIX
                                                              // arg6:REST

#define PRIME_PREEDIT_CONVERT_INPUT  "preedit_convert_input"  // arg1:PATTERN

#define PRIME_GET_ENV                "get_env"                // arg1:KEY
#define PRIME_GET_LABEL              "get_label"              // arg1:PATTERN

#define PRIME_SET_CONTEXT            "set_context"            // CONTEXT
#define PRIME_RESET_CONTEXT          "reset_context"          // noargs

#endif /* __PRIME_COMMANDS_H__ */
